ts/nni_manager/training_service/reusable/gpuScheduler.ts

```diff
@@ -51,46 +53,134 @@ export class GpuScheduler {
+                    gpuIndices: undefined,
+                    environment: undefined,
+                });
+            }
+
+            // Step 2: Allocate Host/GPU for specified trial job
+            // Currenty the requireGPUNum parameter for all trial jobs are identical.
+            if (requiredGPUNum > 0) {
+                // Trial job requires GPU
+                const result: GpuScheduleResult | undefined = this.scheduleGPUHost(environments, requiredGPUNum, trialDetail);
+                if (result !== undefined) {
+                    return result;
+                }
+            } else {
+                // Trail job does not need GPU
+                const allocatedRm: EnvironmentInformation = this.selectMachine(environments, environments);
+
+                return this.allocateHost(requiredGPUNum, allocatedRm, [], trialDetail);
+            }
+
+            return {
+                resultType: ScheduleResultType.TMP_NO_AVAILABLE_GPU,
                 gpuIndices: undefined,
                 environment: undefined,
-            });
-        }
+            };
+        } else {
+            assert(constraint.type === 'Device')
+            if (constraint.gpus.length == 0) {
+                throw new Error("Device constraint is used but no device is specified.");
+            }
+            const gpus = constraint.gpus as Array<[string, number]>;
+            const selectedHost = gpus[0][0];
 
-        // Step 2: Allocate Host/GPU for specified trial job
-        // Currenty the requireGPUNum parameter for all trial jobs are identical.
-        if (requiredGPUNum > 0) {
-            // Trial job requires GPU
-            const result: GpuScheduleResult | undefined = this.scheduleGPUHost(environments, requiredGPUNum, trialDetail);
-            if (result !== undefined) {
-                return result;
+            const hostsOfConstraint: Array<[string, number]> = gpus.filter((gpuTuple: [string, number]) => gpuTuple[0] === selectedHost);
+            if (hostsOfConstraint.length > 1) {
+                throw new Error("Device constraint does not support using multiple hosts")
             }
-        } else {
-            // Trail job does not need GPU
-            const allocatedRm: EnvironmentInformation = this.selectMachine(environments, environments);
+            if (environments.length == 0) {
+                return {
+                    resultType: ScheduleResultType.TMP_NO_AVAILABLE_GPU,
+                    gpuIndices: undefined,
+                    environment: undefined,
+                };
+            }
+            const eligibleEnvironments: EnvironmentInformation[] = environments.filter(
+                (environment: EnvironmentInformation) =>
+                    (environment as RemoteMachineEnvironmentInformation).rmMachineMeta != undefined &&
+                    (environment as RemoteMachineEnvironmentInformation).rmMachineMeta?.host == selectedHost);
```

Comment:
PlacementConstraint is specified for RemoteMachineEnvironmentInformation?

Reply:
Currently yes, because only `RemoteMachineEnvironmentInformation` has the `host` attribute for locating in 
`PlacementConstraint`.

The current `id`, `envId`, and  `name` in `EnvironmentInformation` are generated each time an environment is requested in `TrialDispatcher.requestEnvironment`. So the list of environments could be changed during execution. If we use `envId` for locating environment in `PlacementConstraint`, we need to update the environment list from typescript to python.